File: macos/Sources/Pomodoro/Core/WhistlerConfig.swift
import Foundation

/// Where Whistler's configuration comes from: the non-secret parts from a
/// JSON file beside the history, the secret parts from the Keychain. Assembled
/// into the flat dictionary the Python bridges already read, so only the
/// source changed — and the secrets now reach the bridges through the child
/// process environment rather than through a file on disk.
enum WhistlerConfig {
    static let defaultApiUrl = "https://whistler.nashatech.com"
    static let defaultModel = "openai/gpt-4o-mini"
    static let defaultCalendar = "primary"

    /// Carried over only for a setup migrated before it had a token.
    static let legacyPassword = "whistler-password"

    struct Settings: Equatable {
        var apiUrl: String = WhistlerConfig.defaultApiUrl
        var email: String = ""
        var calendarId: String = WhistlerConfig.defaultCalendar
        var model: String = WhistlerConfig.defaultModel
    }

    private static var settingsURL: URL { DataPaths.whistlerAccount }

    static func readSettings() -> Settings {
        migrate()
        guard let data = try? Data(contentsOf: settingsURL),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return Settings()
        }
        func text(_ key: String, _ fallback: String) -> String {
            let value = json[key] as? String ?? ""
            return value.isEmpty ? fallback : value
        }
        return Settings(
            apiUrl: text("apiUrl", defaultApiUrl),
            email: text("email", ""),
            calendarId: text("calendarId", defaultCalendar),
            model: text("model", defaultModel))
    }

    static func writeSettings(_ settings: Settings) {
        DataPaths.ensureDirectory()
        let json: [String: Any] = [
            "version": 1,
            "apiUrl": settings.apiUrl.hasSuffix("/") ? String(settings.apiUrl.dropLast()) : settings.apiUrl,
            "email": settings.email,
            "calendarId": settings.calendarId,
            "model": settings.model
        ]
        guard let data = try? JSONSerialization.data(
            withJSONObject: json, options: [.prettyPrinted, .sortedKeys]) else { return }
        try? data.write(to: settingsURL, options: .atomic)
    }

    /// The flat shape the bridges read, secrets included. Only ever passed
    /// through a child process environment, never written out.
    static func resolve() -> [String: String] {
        let settings = readSettings()
        var values = [
            "WHISTLER_API_URL": settings.apiUrl,
            "WHISTLER_EMAIL": settings.email,
            "GOOGLE_CALENDAR_ID": settings.calendarId,
            "CALENDAR_ID": settings.calendarId,
            "OPENROUTER_MODEL": settings.model,
            "WHISTLER_SESSION_TOKEN": SecretStore.read(SecretStore.whistlerSession) ?? "",
            "OPENROUTER_API_KEY": SecretStore.read(SecretStore.openRouterKey) ?? ""
        ]
        if let password = SecretStore.read(legacyPassword), !password.isEmpty {
            values["WHISTLER_PASSWORD"] = password
        }
        return values.filter { !$0.value.isEmpty }
    }

    /// Stores what the sign-in produced. The password is never written.
    static func save(_ settings: Settings, sessionToken: String, openRouterKey: String) {
        writeSettings(settings)
        SecretStore.write(SecretStore.whistlerSession, sessionToken)
        SecretStore.write(SecretStore.openRouterKey, openRouterKey)
        SecretStore.delete(legacyPassword)
    }

    /// Moves a pomodoro-whistler.env into the Keychain, then scrubs and
    /// deletes it. A password is carried over only when there is no token to
    /// carry instead, so a working setup keeps working until the next sign-in
    /// replaces it.
    private static func migrate() {
        let legacy = DataPaths.whistlerConfig
        guard FileManager.default.fileExists(atPath: legacy.path) else { return }

        let env = readEnv(legacy)
        if !env.isEmpty {
            writeSettings(Settings(
                apiUrl: env["WHISTLER_API_URL"] ?? defaultApiUrl,
                email: env["WHISTLER_EMAIL"] ?? "",
                calendarId: env["GOOGLE_CALENDAR_ID"] ?? env["CALENDAR_ID"] ?? defaultCalendar,
                model: env["OPENROUTER_MODEL"] ?? defaultModel))

            SecretStore.write(SecretStore.openRouterKey, env["OPENROUTER_API_KEY"] ?? "")
            let token = env["WHISTLER_SESSION_TOKEN"] ?? ""
            if token.isEmpty {
                SecretStore.write(legacyPassword, env["WHISTLER_PASSWORD"] ?? "")
            } else {
                SecretStore.write(SecretStore.whistlerSession, token)
            }
        }

        // Overwrite before unlinking: deleting a file leaves its contents on
        // the disk, and this one held a password.
        if let size = try? FileManager.default.attributesOfItem(atPath: legacy.path)[.size] as? Int {
            try? Data(count: size).write(to: legacy)
        }
        try? FileManager.default.removeItem(at: legacy)
    }

    /// The legacy file's format, parsed here so migration does not depend on
    /// a main-actor-isolated helper.
    private static func readEnv(_ url: URL) -> [String: String] {
        guard let text = try? String(contentsOf: url, encoding: .utf8) else { return [:] }
        var values: [String: String] = [:]
        for line in text.split(separator: "\n") {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            guard !trimmed.isEmpty, !trimmed.hasPrefix("#"),
                  let separator = trimmed.firstIndex(of: "=") else { continue }
            let key = String(trimmed[trimmed.startIndex..<separator]).trimmingCharacters(in: .whitespaces)
            var value = String(trimmed[trimmed.index(after: separator)...])
                .trimmingCharacters(in: .whitespaces)
            if value.count >= 2, let first = value.first, let last = value.last,
               first == last, first == "\"" || first == "'" {
                value = String(value.dropFirst().dropLast())
            }
            values[key] = value
        }
        return values
    }
}

File: windows/src/Pomodoro.Core/WhistlerConfig.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Pomodoro.Core;

/// <summary>
/// Where Whistler's configuration comes from: the non-secret parts from a
/// JSON file beside the history, the secret parts from the OS credential
/// store. Assembled into the flat dictionary the importer and the two API
/// clients already expect, so only the source changed, not the consumers.
///
/// The old pomodoro-whistler.env held everything in plain text, including the
/// account password typed there to obtain a session token. It is migrated on
/// first read and then removed.
/// </summary>
public static class WhistlerConfig
{
    public const string DefaultApiUrl = "https://whistler.nashatech.com";
    public const string DefaultModel = "openai/gpt-4o-mini";
    public const string DefaultCalendar = "primary";

    /// <summary>Kept only until a session token replaces it; see Migrate.</summary>
    public const string LegacyPassword = "whistler-password";

    private static string SettingsPath => DataPaths.WhistlerAccount;

    public sealed record Settings
    {
        public string ApiUrl { get; init; } = DefaultApiUrl;
        public string Email { get; init; } = string.Empty;
        public string CalendarId { get; init; } = DefaultCalendar;
        public string Model { get; init; } = DefaultModel;
    }

    public static Settings ReadSettings()
    {
        Migrate();
        var text = AtomicFile.Read(SettingsPath);
        if (text is null) return new Settings();
        try
        {
            if (JsonNode.Parse(text) is not JsonObject json) return new Settings();
            return new Settings
            {
                ApiUrl = Text(json["apiUrl"], DefaultApiUrl),
                Email = Text(json["email"], string.Empty),
                CalendarId = Text(json["calendarId"], DefaultCalendar),
                Model = Text(json["model"], DefaultModel)
            };
        }
        catch (JsonException)
        {
            return new Settings();
        }
    }

    public static void WriteSettings(Settings settings)
    {
        DataPaths.EnsureDirectory();
        AtomicFile.Write(SettingsPath, Persistence.Json(new JsonObject
        {
            ["version"] = 1,
            ["apiUrl"] = settings.ApiUrl.TrimEnd('/'),
            ["email"] = settings.Email,
            ["calendarId"] = settings.CalendarId,
            ["model"] = settings.Model
        }));
    }

    /// <summary>The flat shape the importer and the API clients read.</summary>
    public static Dictionary<string, string> Resolve()
    {
        var settings = ReadSettings();
        var values = new Dictionary<string, string>(StringComparer.Ordinal)
        {
            ["WHISTLER_API_URL"] = settings.ApiUrl,
            ["WHISTLER_EMAIL"] = settings.Email,
            ["GOOGLE_CALENDAR_ID"] = settings.CalendarId,
            ["OPENROUTER_MODEL"] = settings.Model,
            ["WHISTLER_SESSION_TOKEN"] = SecretStore.Read(SecretStore.WhistlerSession) ?? string.Empty,
            ["OPENROUTER_API_KEY"] = SecretStore.Read(SecretStore.OpenRouterKey) ?? string.Empty
        };
        // Only present for a configuration migrated before it had a token.
        var password = SecretStore.Read(LegacyPassword);
        if (!string.IsNullOrEmpty(password)) values["WHISTLER_PASSWORD"] = password;
        return values;
    }

    /// <summary>True when there is enough to reach Whistler and OpenRouter.</summary>
    public static bool IsConfigured()
    {
        if (!SecretStore.Has(SecretStore.OpenRouterKey) &&
            string.IsNullOrEmpty(Environment.GetEnvironmentVariable("OPENROUTER_API_KEY")))
        {
            return false;
        }
        return SecretStore.Has(SecretStore.WhistlerSession) || SecretStore.Has(LegacyPassword);
    }

    /// <summary>Stores what the sign-in produced. The password is never written.</summary>
    public static void Save(Settings settings, string sessionToken, string openRouterKey)
    {
        WriteSettings(settings);
        SecretStore.Write(SecretStore.WhistlerSession, sessionToken);
        SecretStore.Write(SecretStore.OpenRouterKey, openRouterKey);
        // A fresh token makes any migrated password redundant.
        SecretStore.Delete(LegacyPassword);
    }

    public static void Clear()
    {
        SecretStore.Delete(SecretStore.WhistlerSession);
        SecretStore.Delete(SecretStore.OpenRouterKey);
        SecretStore.Delete(LegacyPassword);
        try { File.Delete(SettingsPath); } catch (IOException) { /* already gone */ }
    }

    /// <summary>
    /// Moves a pomodoro-whistler.env into the credential store, then scrubs
    /// and deletes it. A password is carried over only when there is no token
    /// to carry instead, so a working setup keeps working; the next sign-in
    /// replaces it.
    /// </summary>
    private static void Migrate()
    {
        var legacy = DataPaths.WhistlerConfig;
        if (!File.Exists(legacy)) return;

        try
        {
            var env = IntegrationStatus.ReadEnv(legacy);
            if (env.Count > 0)
            {
                WriteSettings(new Settings
                {
                    ApiUrl = env.GetValueOrDefault("WHISTLER_API_URL", DefaultApiUrl),
                    Email = env.GetValueOrDefault("WHISTLER_EMAIL", string.Empty),
                    CalendarId = env.GetValueOrDefault("GOOGLE_CALENDAR_ID", DefaultCalendar),
                    Model = env.GetValueOrDefault("OPENROUTER_MODEL", DefaultModel)
                });

                var token = env.GetValueOrDefault("WHISTLER_SESSION_TOKEN", string.Empty);
                SecretStore.Write(SecretStore.OpenRouterKey,
                    env.GetValueOrDefault("OPENROUTER_API_KEY", string.Empty));
                if (token.Length > 0)
                {
                    SecretStore.Write(SecretStore.WhistlerSession, token);
                }
                else
                {
                    SecretStore.Write(LegacyPassword,
                        env.GetValueOrDefault("WHISTLER_PASSWORD", string.Empty));
                }
            }

            // Overwrite before unlinking: deleting a file leaves its contents
            // on the disk, and this one held a password.
            var length = new FileInfo(legacy).Length;
            File.WriteAllBytes(legacy, new byte[length]);
            File.Delete(legacy);
        }
        catch (Exception error) when (error is IOException or UnauthorizedAccessException)
        {
            // Leave it in place; the next launch tries again.
        }
    }

    private static string Text(JsonNode? node, string fallback)
    {
        var value = node?.GetValue<string>();
        return string.IsNullOrEmpty(value) ? fallback : value;
    }
}
